java/core/src/test/org/apache/orc/TestOrcTimestampPPD.java

```diff
@@ -0,0 +1,173 @@
+      }
+    }
+    rows.close();
+    ColumnStatistics[] colStats = reader.getStatistics();
+    Timestamp gotMin = ((TimestampColumnStatistics) colStats[0]).getMinimum();
+    assertEquals("1970-01-01 00:00:00.0", gotMin.toString());
+
+    Timestamp gotMax = ((TimestampColumnStatistics) colStats[0]).getMaximum();
+    assertEquals("1970-01-01 00:00:00.0", gotMax.toString());
+
+    // Make sure PPD is now passing
+    Assert.assertEquals(SearchArgument.TruthValue.YES_NO, RecordReaderImpl.evaluatePredicate(colStats[0],
+        SearchArgumentFactory.newBuilder()
+            .equals("c", PredicateLeaf.Type.TIMESTAMP, Timestamp.valueOf("1970-01-01 00:00:00.0005")).build()
+            .getLeaves().get(0), null));
+  }
+
+  @Test
+  public void testSubMsComplexStats() throws IOException {
+    TypeDescription schema = TypeDescription.createTimestamp();
+    TimeZone.setDefault(TimeZone.getTimeZone("America/Los_Angeles"));
+
+    Writer writer = OrcFile.createWriter(testFilePath,
+        OrcFile.writerOptions(conf).setSchema(schema).stripeSize(100000).bufferSize(10000)
+            .version(OrcFile.Version.CURRENT));
+
+    List<Timestamp> tslist = Lists.newArrayList();
+    tslist.add(Timestamp.valueOf("2037-01-01 00:00:00.002009"));
+    tslist.add(Timestamp.valueOf("2037-01-01 00:00:00.002079"));
+    tslist.add(Timestamp.valueOf("2037-01-01 00:00:00.002099"));
+    tslist.add(Timestamp.valueOf("2037-01-01 00:00:00.002199"));
+    tslist.add(Timestamp.valueOf("2037-01-01 00:00:00.002699"));
+    tslist.add(Timestamp.valueOf("2037-01-01 00:00:00.002999"));
+
+    VectorizedRowBatch batch = schema.createRowBatch();
+    TimestampColumnVector times = (TimestampColumnVector) batch.cols[0];
+    for (Timestamp t : tslist) {
+      times.set(batch.size++, t);
+    }
+    times.isRepeating = true;
+    writer.addRowBatch(batch);
+    // Done writing to file
+    writer.close();
+
+    TimeZone.setDefault(TimeZone.getTimeZone("America/Los_Angeles"));
+    // Now reading
+    Reader reader = OrcFile.createReader(testFilePath, OrcFile.readerOptions(conf).filesystem(fs));
+
+    RecordReader rows = reader.rows();
+    batch = reader.getSchema().createRowBatch();
+    times = (TimestampColumnVector) batch.cols[0];
+    while (rows.nextBatch(batch)) {
+      for (int r = 0; r < batch.size; ++r) {
+        Assert.assertEquals(tslist.get(0), times.asScratchTimestamp(r));
+        Assert.assertEquals(tslist.get(0).getNanos(), times.asScratchTimestamp(r).getNanos());
+      }
+    }
+    rows.close();
+    ColumnStatistics[] colStats = reader.getStatistics();
+
+    // Make sure PPD is now passing
+    Assert.assertEquals(SearchArgument.TruthValue.YES_NO, RecordReaderImpl.evaluatePredicate(colStats[0],
+        SearchArgumentFactory.newBuilder()
+            .equals("c", PredicateLeaf.Type.TIMESTAMP, Timestamp.valueOf("2037-01-01 00:00:00.001009")).build()
+            .getLeaves().get(0), null));
+  }
```

Comment:
Please add a test with a timestamp equality and both approximate and precise min/max.